### PR TITLE
[New Provider Function]: `arn_parse` and `arn_build`

### PR DESCRIPTION
**Pull request by jar-b**
This PR is a work in-progress. Remaining tasks include:

- [x] Dependencies all point to proper versions.
- [x] Contributor guide updated to describe cases for provider functions, document best practices.
- [x] Registry documentation.
- [x] Resolving (or skipping) CI checks requiring Terraform 1.8+ to succeed (ie. `terrafmt`, functions unit tests)

---

### Description

The `arn_parse` and `arn_build` provider functions will allow practioners to handle ARN (Amazon Resource Name) manipulation with built-in provider functions. The `arn_parse` function provides similar utility to the `aws_arn` data source, but with the benefit of running earlier in the execution order. 

### Relations

- Relates https://github.com/hashicorp/terraform-plugin-go/pull/351
- Relates https://github.com/hashicorp/terraform-plugin-framework/pull/889
- Relates https://github.com/hashicorp/terraform-plugin-mux/pull/209
- Relates https://github.com/hashicorp/terraform-plugin-sdk/pull/1288
- Relates https://github.com/hashicorp/terraform/pull/34383

### Output from Acceptance Testing

```console
% TF_ACC_TERRAFORM_PATH=/Users/jaredbaker/go/bin/terraform go test -count=1 -run='^TestARN' -v ./internal/function/
=== RUN   TestARNBuildFunction_known
=== PAUSE TestARNBuildFunction_known
=== RUN   TestARNParseFunction_known
=== PAUSE TestARNParseFunction_known
=== RUN   TestARNParseFunction_invalid
=== PAUSE TestARNParseFunction_invalid
=== CONT  TestARNBuildFunction_known
=== CONT  TestARNParseFunction_invalid
=== CONT  TestARNParseFunction_known
--- PASS: TestARNParseFunction_invalid (1.01s)
--- PASS: TestARNParseFunction_known (9.21s)
--- PASS: TestARNBuildFunction_known (9.25s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/function   12.461s
```

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -14,6 +14,7 @@ nav:
           - Resource: add-a-new-resource.md
           - Service: add-a-new-service.md
           - Data source: add-a-new-datasource.md
+          - Function: add-a-new-function.md
```
Comment (YakDriver) [inline mkdocs.yml:17]: Do we need a little more than just "Function"? Provider function? Not sure.

Comment (jar-b) [reply mkdocs.yml:17]: I went with `Function` to align with how the Terraform Plugin Framework package is named, and how its used in the documentation:
https://developer.hashicorp.com/terraform/plugin/framework/functions

"Provider-defined" function is another term that has been used, and though I think the location of this document (the contributing docs for a provider) implies that namespace, I don't have a strong opinion either way.

**Comment by jar-b** [comment]
With Terraform `1.8.0-beta1`:

```console
% go test -count=1 -v ./internal/function/...

--- PASS: TestARNParseFunction_invalid (0.88s)
--- PASS: TestARNBuildFunction_known (4.91s)
--- PASS: TestARNParseFunction_known (4.91s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/function   9.017s
```

With Terraform `1.7.X`:

```console
% go test -count=1 -v ./internal/function/...

    arn_parse_function_test.go:20: Terraform CLI version 1.7.4 is below minimum version 1.8.0-beta1: skipping test
--- SKIP: TestARNParseFunction_known (0.46s)
=== NAME  TestARNBuildFunction_known
    arn_build_function_test.go:18: Terraform CLI version 1.7.4 is below minimum version 1.8.0-beta1: skipping test
--- SKIP: TestARNBuildFunction_known (0.47s)
=== NAME  TestARNParseFunction_invalid
    arn_parse_function_test.go:40: Terraform CLI version 1.7.4 is below minimum version 1.8.0-beta1: skipping test
--- SKIP: TestARNParseFunction_invalid (0.48s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/function   4.578s
```

**Comment by ewbankkit** [comment]
Not a blocker: We need to add a `new-function` GitHub label.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% go test -count=1 -v ./internal/function/...
=== RUN   TestARNBuildFunction_known
=== PAUSE TestARNBuildFunction_known
=== RUN   TestARNParseFunction_known
=== PAUSE TestARNParseFunction_known
=== RUN   TestARNParseFunction_invalid
=== PAUSE TestARNParseFunction_invalid
=== CONT  TestARNBuildFunction_known
=== CONT  TestARNParseFunction_known
=== CONT  TestARNParseFunction_invalid
    arn_parse_function_test.go:40: Terraform CLI version 1.0.11 is below minimum version 1.8.0-beta1: skipping test
=== NAME  TestARNBuildFunction_known
    arn_build_function_test.go:18: Terraform CLI version 1.0.11 is below minimum version 1.8.0-beta1: skipping test
=== NAME  TestARNParseFunction_known
    arn_parse_function_test.go:20: Terraform CLI version 1.0.11 is below minimum version 1.8.0-beta1: skipping test
--- SKIP: TestARNParseFunction_invalid (0.50s)
--- SKIP: TestARNBuildFunction_known (0.50s)
--- SKIP: TestARNParseFunction_known (0.50s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/function	7.893s
```
```console
% go test -count=1 -v ./internal/function/...
=== RUN   TestARNBuildFunction_known
=== PAUSE TestARNBuildFunction_known
=== RUN   TestARNParseFunction_known
=== PAUSE TestARNParseFunction_known
=== RUN   TestARNParseFunction_invalid
=== PAUSE TestARNParseFunction_invalid
=== CONT  TestARNBuildFunction_known
=== CONT  TestARNParseFunction_invalid
=== CONT  TestARNParseFunction_known
--- PASS: TestARNParseFunction_invalid (1.70s)
--- PASS: TestARNBuildFunction_known (8.16s)
--- PASS: TestARNParseFunction_known (8.22s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/function	15.317s
```

**Pull request merged**